File: threshold_frost/src/signing.go
package threshold_signing

import (
	thres "github.com/ArkLabsHQ/thresholdmagic/thresholdcore"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
)

type SigningPackage struct {
	Commitments map[thres.Identifier]SigningCommitments
	Message     []byte
}

func (s *SigningPackage) SigningCommitment(id thres.Identifier) (SigningCommitments, bool) {
	c, ok := s.Commitments[id]
	return c, ok
}

func (s *SigningPackage) encodeGroupCommitmentList() []byte {
	groupCommitmentList := s.Commitments

	var buf []byte
	for id, comm := range groupCommitmentList {
		idSerialised := id.Serialize()

		buf = append(buf, idSerialised...)

		hidingBytes := SerializePointCompressed(comm.Hiding)
		buf = append(buf, hidingBytes...)

		bindingBytes := SerializePointCompressed(comm.Binding)
		buf = append(buf, bindingBytes...)
	}

	return buf
}

func (s *SigningPackage) bindingFactorPreimages(
	vk thres.VerifyingKey) ([]BindingFactorPreimage, error) {
	var prefix []byte

	var affinePoint secp256k1.JacobianPoint
	affinePoint.Set(&vk.E)
	affinePoint.ToAffine()

	vkBytes := secp256k1.NewPublicKey(&affinePoint.X, &affinePoint.Y).SerializeCompressed()

	prefix = append(prefix, vkBytes...)

	// H4(message)
	h4 := H4(s.Message)
	prefix = append(prefix, h4...)

	// H5(encode_group_commitments(commitments))
	encGC := s.encodeGroupCommitmentList()
	h5 := H5(encGC)
	prefix = append(prefix, h5...)

	// build per-identifier preimages
	out := make([]BindingFactorPreimage, 0, len(s.Commitments))

	for id := range s.Commitments {
		buf := make([]byte, 0, len(prefix)+len(id.Serialize()))
		buf = append(buf, prefix...)
		buf = append(buf, id.Serialize()...)
		out = append(out, BindingFactorPreimage{
			ID:       id,
			Preimage: buf,
		})
	}
	return out, nil
}

// / Performed once by each participant selected for the signing operation.
// /
// / Implements [`sign`] from the spec.
// /
// / Receives the message to be signed and a set of signing commitments and a set
// / of randomizing commitments to be used in that signing operation, including
// / that for this participant.
// /
// / Assumes the participant has already determined which nonce corresponds with
// / the commitment that was assigned by the coordinator in the SigningPackage.
// /
// / [`sign`]: https://datatracker.ietf.org/doc/html/rfc9591#name-round-two-signature-share-g
func Sign(
	signingPackage SigningPackage, signingNonce SigningNonce, keyPackage thres.KeyPackage) (*SignatureShare, error) {

	if len(signingPackage.Commitments) < int(keyPackage.MinSigners) {
		// Not enough participants to sign
		return nil, thres.ErrIncorrectNumberOfCommitments
	}

	commitment := signingPackage.Commitments[keyPackage.Identifier]

	if signingNonce.commitments != commitment {
		// Nonce does not match commitment
		return nil, ErrInvalidCommitment
	}

	bindingFactorList, err := ComputeBindingFactorList(&signingPackage, keyPackage.VerifyingKey)
	if err != nil {
		return nil, ErrIncorrectBindingFactorPreimages
	}

	groupCommitment, err := ComputeGroupCommitment(&signingPackage, bindingFactorList)

	lambda_i := DeriveInterpolatingValue(keyPackage.Identifier, &signingPackage)

	challenge, err := ComputeChallenge(&groupCommitment.elem, keyPackage.VerifyingKey, signingPackage.Message)
	if err != nil {
		return nil, err
	}

	signatureShare := ComputeSignatureShare(signingNonce, bindingFactorList.F[keyPackage.Identifier], *lambda_i, keyPackage, *challenge)

	return &signatureShare, nil
}

func ComputeChallenge(R *secp256k1.JacobianPoint, vk thres.VerifyingKey, message []byte) (*secp256k1.ModNScalar, error) {
	// H2(encode_point(R) || encode_point(Y) || message)
	var affineR secp256k1.JacobianPoint
	affineR.Set(R)
	affineR.ToAffine()
	RBytes := secp256k1.NewPublicKey(&affineR.X, &affineR.Y).SerializeCompressed()

	var affineY secp256k1.JacobianPoint
	affineY.Set(&vk.E)
	affineY.ToAffine()
	YBytes := secp256k1.NewPublicKey(&affineY.X, &affineY.Y).SerializeCompressed()

	prefix := append(RBytes, YBytes...)
	prefix = append(prefix, message...)

	e := H2(prefix)
	return &e, nil
}

// Aggregate sums signature shares into a single Schnorr signature and verifies it
// against the group public key. Optionally, you can do cheater-detection by
// verifying each share if the final signature verification fails.
func Aggregate(
	signingPackage *SigningPackage,
	signatureShares map[thres.Identifier]SignatureShare, // key = string(id.Serialize())
	pubkeys thres.PublicKeyPackage,
) (*thres.Signature, error) {
	// 1) identifier set must match
	if len(signingPackage.Commitments) != len(signatureShares) {
		return nil, ErrUnknownIdentifier
	}
	for id, _ := range signingPackage.Commitments {
		if _, ok := signatureShares[id]; !ok {
			return nil, ErrUnknownIdentifier
		}

		if _, ok := pubkeys.VerifyingShares[id]; !ok {
			return nil, ErrUnknownIdentifier
		}
	}

	bfl, err := ComputeBindingFactorList(signingPackage, pubkeys.VerifyingKey)
	if err != nil {
		return nil, err
	}
	groupCommitment, err := ComputeGroupCommitment(signingPackage, bfl)
	if err != nil {
		return nil, err
	}

	// aggregate z = sum(z_i)
	var z secp256k1.ModNScalar // zero by default
	for _, sh := range signatureShares {
		// z = z + z_i (mod n); ModNScalar.Add mutates receiver and returns it
		z.Add(&sh.s)
	}

	sig := &thres.Signature{
		R: groupCommitment.elem,
		Z: z,
	}

	//verify final signature with the group verifying key
	if pubkeys.VerifyingKey.Verify(signingPackage.Message, *sig) {

		if err2 := DetectCheater(&groupCommitment, &pubkeys, signingPackage, signatureShares, bfl); err2 != nil {
			return nil, err2
		}
		return sig, err
	}

	return sig, ErrorWrongSignature
}

func DetectCheater(
	groupCommitment *GroupCommitment,
	pubkeys *thres.PublicKeyPackage,
	signingPackage *SigningPackage,
	signatureShares map[thres.Identifier]SignatureShare,
	bfl BindingFactorList,
) error {
	// Compute per-message challenge
	e, err := ComputeChallenge(&groupCommitment.elem, pubkeys.VerifyingKey, signingPackage.Message)
	if err != nil {
		return err
	}

	// Verify each share
	for id, sshare := range signatureShares {
		verShare, ok := pubkeys.VerifyingShares[id]
		if !ok {
			return ErrUnknownIdentifier
		}

		if err := verifySignatureSharePrecomputed(
			id,
			signingPackage,
			bfl,
			groupCommitment,
			&sshare,
			&verShare,
			e,
		); err != nil {
			return err
		}
	}

	// If all passed, fall back to a generic error to mirror Rust's "should not reach"
	return ErrorInvalidSignature
}

func VerifySignatureShare(
	identifier thres.Identifier,
	verifyingShare *secp256k1.JacobianPoint,
	signatureShare *SignatureShare,
	signingPackage *SigningPackage,
	verifyingKey thres.VerifyingKey,
) error {
	// Binding factors and group commitment
	bfl, err := ComputeBindingFactorList(signingPackage, verifyingKey)
	if err != nil {
		return err
	}
	groupCommitment, err := ComputeGroupCommitment(signingPackage, bfl)
	if err != nil {
		return err
	}

	// Challenge
	e, err := ComputeChallenge(&groupCommitment.elem, verifyingKey, signingPackage.Message)
	if err != nil {
		return err
	}

	// Final check
	return verifySignatureSharePrecomputed(
		identifier,
		signingPackage,
		bfl,
		&groupCommitment,
		signatureShare,
		verifyingShare,
		e,
	)
}

// verifySignatureSharePrecomputed performs the share check with precomputed values.
func verifySignatureSharePrecomputed(
	signatureShareIdentifier thres.Identifier,
	signingPackage *SigningPackage,
	bfl BindingFactorList,
	groupCommitment *GroupCommitment,
	signatureShare *SignatureShare,
	verifyingShare *secp256k1.JacobianPoint,
	challenge *secp256k1.ModNScalar,
) error {

	ids := make([]thres.Identifier, 0, len(signingPackage.Commitments))
	for k := range signingPackage.Commitments {
		ids = append(ids, k)
	}
	// λ_i
	lambdaI := thres.LagrangeCoeffAtZero(signatureShareIdentifier, ids)

	// ρ_i for this identifier
	bf, ok := bfl.Get(signatureShareIdentifier)
	if !ok {
		return ErrUnknownIdentifier
	}

	// commitments for this signer
	comm, ok := signingPackage.Commitments[signatureShareIdentifier]
	if !ok {
		return ErrUnknownIdentifier
	}

	// R_i share for this signer
	Rshare := comm.toGroupCommitmentShare(bf.Scalar)

	// Relation check
	res := signatureShare.verify(
		signatureShareIdentifier,
		*verifyingShare,
		*lambdaI,
		*challenge,
		Rshare,
	)

	if !res {
		return ErrorInvalidSignature
	}

	return nil
}

File: threshold/lib/frost/signing.dart
import 'dart:typed_data';

import 'package:pointycastle/ecc/api.dart';
import 'package:threshold/core/dkg.dart'; // For KeyPackage, PublicKeyPackage, Signature
import 'package:threshold/core/identifier.dart';
import 'package:threshold/core/share.dart';
import 'package:threshold/core/utils.dart'; // for constants, BigInt utils
import 'package:threshold/frost/errors.dart';
import 'package:threshold/frost/hasher.dart';
import 'package:threshold/frost/utils.dart';
import 'package:threshold/frost/commitment.dart';
import 'package:threshold/frost/binding.dart';

class SignatureShare {
  final BigInt s;
  SignatureShare(this.s);
}

// Sign implements [sign] from the spec.
// Returns a signature share.
SignatureShare sign(
  SigningPackage signingPackage,
  SigningNonce signingNonce,
  KeyPackage keyPackage,
) {
  if (signingPackage.commitments.length < keyPackage.minSigners) {
    throw errIncorrectNumberOfCommitments;
  }

  final commitment = signingPackage.commitments[keyPackage.identifier];
  if (commitment == null) {
    // This participant is not in the signing set
    throw errInvalidCommitment; // Using similar error
  }

  // Check if nonce matches commitment.
  // In our Dart implementation, we might not have '==' on objects easily unless overridden.
  // We check byte equivalence of points.
  if (!pointsEqual(signingNonce.commitments.binding, commitment.binding) ||
      !pointsEqual(signingNonce.commitments.hiding, commitment.hiding)) {
    throw errInvalidCommitment;
  }

  final bfl = computeBindingFactorList(signingPackage, keyPackage.verifyingKey);
  final groupCommitment = computeGroupCommitment(signingPackage, bfl);

  final lambdaI = deriveInterpolatingValue(
    keyPackage.identifier,
    signingPackage,
  );
  final challenge = computeChallenge(
    groupCommitment.elem,
    keyPackage.verifyingKey,
    signingPackage.message,
  );

  // Compute Signature Share
  // z_i = d_i + (e_i * rho_i) + lambda_i * s_i * c
  final bf = bfl.get(keyPackage.identifier);
  if (bf == null) throw errIncorrectBindingFactorPreimages; // Should not happen

  return computeSignatureShare(
    signingNonce,
    bf.scalar,
    lambdaI,
    keyPackage,
    challenge,
  );
}

BigInt computeChallenge(ECPoint R, VerifyingKey vk, Uint8List message) {
  final RBytes = serializePointCompressed(R);
  final YBytes = serializePointCompressed(vk.E);

  final builder = BytesBuilder();
  builder.add(RBytes);
  builder.add(YBytes);
  builder.add(message);

  return h2(builder.toBytes());
}

BigInt deriveInterpolatingValue(Identifier id, SigningPackage pkg) {
  final ids = sortedCommitmentIDs(pkg.commitments.keys.toList());
  return lagrangeCoeffAtZero(id, ids);
}

// Helper to compute signature share
SignatureShare computeSignatureShare(
  SigningNonce nonce,
  BigInt rhoI,
  BigInt lambdaI,
  KeyPackage keyPackage,
  BigInt challenge,
) {
  // z_i = d_i + (e_i * rho_i) + lambda_i * s_i * c
  // Note: In `signing.go`: ComputeSignatureShare seems to be missing in the file provided?
  // Wait, line 118 calls `ComputeSignatureShare`.
  // It was likely in another file or implicitly defined?
  // Ah, the view of `signing.go` line 118 calls it. But `signing.go` does not DEFINE it.
  // It must be in `share.go` or somewhere else within `threshold_signing`.
  // I need to implement what FROST spec says.

  // d_i = hiding nonce scalar
  // e_i = binding nonce scalar
  // rho_i = binding factor
  // lambda_i = lagrange coeff
  // s_i = long term secret share
  // c = challenge

  // z = d + (e * rho) + lambda * s * c  (mod q)

  final d = nonce.hiding;
  final e = nonce.binding;
  final s = keyPackage.secretShare;
  final c = challenge;

  final modulus = secp256k1Curve.n;

  final eRho = (e * rhoI) % modulus;
  final lsc = (lambdaI * s * c) % modulus;

  final z = (d + eRho + lsc) % modulus;

  return SignatureShare(z);
}

// Aggregate
// Returns Signature (R, z)
Signature aggregate(
  SigningPackage signingPackage,
  Map<Identifier, SignatureShare> signatureShares,
  PublicKeyPackage pubkeys,
) {
  // 1. Check identifiers
  if (signingPackage.commitments.length != signatureShares.length) {
    throw errUnknownIdentifier;
  }
  for (final id in signingPackage.commitments.keys) {
    if (!signatureShares.containsKey(id) ||
        !pubkeys.verifyingShares.containsKey(id)) {
      throw errUnknownIdentifier;
    }
  }

  final bfl = computeBindingFactorList(signingPackage, pubkeys.verifyingKey);
  final groupCommitment = computeGroupCommitment(signingPackage, bfl);

  // Aggregate z = sum(z_i)
  var z = BigInt.zero;
  final modulus = secp256k1Curve.n;

  for (final share in signatureShares.values) {
    z = (z + share.s) % modulus;
  }

  final sig = Signature(groupCommitment.elem, z);

  // Verify final signature
  final challenge = computeChallenge(
    groupCommitment.elem,
    pubkeys.verifyingKey,
    signingPackage.message,
  );

  // Verify: z * G == R + c * Y
  final zG = (secp256k1Curve.G * z)!;

  final cY = (pubkeys.verifyingKey.E * challenge)!;
  final R_plus_cY = (groupCommitment.elem + cY)!;

  if (pointsEqual(zG, R_plus_cY)) {
    return sig;
  }

  // Cheater detection would go here (verifySignatureShare)
  throw errorInvalidSignature;
}

// Verify Signature Share
void verifySignatureShare(
  Identifier identifier,
  ECPoint verifyingShare,
  SignatureShare signatureShare,
  SigningPackage signingPackage,
  VerifyingKey verifyingKey,
) {
  // Binding factors and group commitment
  final bfl = computeBindingFactorList(signingPackage, verifyingKey);
  final groupCommitment = computeGroupCommitment(signingPackage, bfl);

  final challenge = computeChallenge(
    groupCommitment.elem,
    verifyingKey,
    signingPackage.message,
  );

  // Verify:
  // z_i * G == R_i + c * lambda_i * Y_i
  // Where R_i = D_i + rho_i * E_i (Wait, check commitment definition)
  // commitment.go: toGroupCommitmentShare: sum = Hiding + bindingScalar * Binding
  // So R_i = H_i + rho_i * B_i

  final comm = signingPackage.commitments[identifier];
  if (comm == null) throw errUnknownIdentifier;

  final bf = bfl.get(identifier);
  if (bf == null) throw errUnknownIdentifier;

  final R_share = comm.toGroupCommitmentShare(bf.scalar).elem; // H + rho*B

  final lambdaI = deriveInterpolatingValue(identifier, signingPackage);

  // LHS: z_i * G
  final LHS = (secp256k1Curve.G * signatureShare.s)!;

  // RHS: R_share + c * lambda_i * Y_i
  final c_lambda = (challenge * lambdaI) % secp256k1Curve.n;
  final term2 = (verifyingShare * c_lambda)!;

  final RHS = (R_share + term2)!;

  if (!pointsEqual(LHS, RHS)) {
    throw errorInvalidSignature;
  }
}

bool pointsEqual(ECPoint a, ECPoint b) {
  if (a.isInfinity && b.isInfinity) return true;
  if (a.isInfinity || b.isInfinity) return false;
  return a.x!.toBigInteger() == b.x!.toBigInteger() &&
      a.y!.toBigInteger() == b.y!.toBigInteger();
}
